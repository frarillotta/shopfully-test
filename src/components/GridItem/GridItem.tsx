import styled from 'styled-components';
import Image from 'next/image';
import { CompletedFlyer } from '../../types';
import { Heart } from '../SVGs/Heart';
import { memo } from 'react';

type GridItemProps = {
	flyer: CompletedFlyer,
	isBookmarked: boolean,
	addFlyer: (flyerId: number) => void,
	removeFlyer: (flyerId: number) => void
};


export const GridItem = memo(({ flyer, isBookmarked, addFlyer, removeFlyer }: GridItemProps) => {
	const { id, title, retailer_name, category_name, asset } = flyer;

	return <Wrapper>
		<NextImage
			src={asset} 
			height={175} 
			width={250}
			alt={'waiting'}
		/>		
		<Description>
			<RetailerName id={'flyer-retailer-name'}>
				{retailer_name}
			</RetailerName>
			<Title id={'flyer-title'}>
				{title}
			</Title>
			<CategoryName id={'flyer-category-name'}>
				{category_name}
			</CategoryName>
			<Bookmarked  data-t={`flyer-is-bookmarked=${isBookmarked}`} isBookmarked={isBookmarked} onClick={() => isBookmarked ? removeFlyer(id) : addFlyer(id)}>
				<Heart size={28}></Heart>
			</Bookmarked>
		</Description>
	</Wrapper>;
});
GridItem.displayName = 'GridItem';

const Wrapper = styled.article`
	display: grid;
	height: 350px;
	width: 100%;
	overflow: hidden;
	grid-template-rows: repeat(2, 1fr);
	grid-template-areas: 
		"picture"
		"description";
	border-radius: 5px;
	box-shadow: 0px 5px 5px -3px rgb(0 0 0 / 20%), 0px 8px 10px 1px rgb(0 0 0 / 14%), 0px 3px 14px 2px rgb(0 0 0 / 12%);

`;

const NextImage = styled(Image)`
	grid-area: picture;
	object-fit: cover;
	object-position: left top;
`;

const Description = styled.div`

	grid-area: description;
	display: flex;
	flex-direction: column;
	padding: 20px;
	gap: 10px;
	justify-content: space-between;
`;

const RetailerName = styled.span`

`;


const Title = styled.span`
	font-size: 1.2rem;
	font-weight: bold;
`;


const CategoryName = styled.span`
	font-size: 0.8rem;
	color: grey;
`;

const Bookmarked = styled.button`
	width: 30px;
	fill: ${props => props.isBookmarked ? 'red' : 'grey'};
	background-color: white;
`;