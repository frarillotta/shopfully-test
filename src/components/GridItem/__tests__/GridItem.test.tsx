import { render } from '@testing-library/react';
import {GridItem} from '../GridItem';

describe('GridItem', () => {

	const mockFlyer = {
		'id': 694041,
		'retailer_name': 'mockName',
		'category_name': 'mockName',
		'title': 'Sport acquatici - sup gonfiabile',
		'retailer_id': 3,
		'start_date': '2021-06-01',
		'end_date': '2021-06-30',
		'is_published': 0,
		'asset': 'https://it-it-media.shopfully.cloud/images/volantini/big_694041.jpg',
		'category_id': 7
	};
	const addFlyer = jest.fn();
	const removeFlyer = jest.fn();
	
	describe('when the grid item is rendered and is bookmarked', () => {
		
		it('should have a consistent snapshot', () => {
			const {container} = render(<GridItem
				addFlyer={addFlyer} flyer={mockFlyer} isBookmarked={true} removeFlyer={removeFlyer}
			/>
			);
			expect(container).toMatchSnapshot();
		});

	});
	describe('when the grid item is rendered and is not bookmarked', () => {
		
		it('should have a consistent snapshot', () => {
			const {container} = render(<GridItem
				addFlyer={addFlyer} flyer={mockFlyer} isBookmarked={false} removeFlyer={removeFlyer}
			/>
			);
			expect(container).toMatchSnapshot();
		});

	});
	
});