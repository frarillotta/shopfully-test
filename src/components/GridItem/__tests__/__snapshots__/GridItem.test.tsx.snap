// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridItem when the grid item is rendered and is bookmarked should have a consistent snapshot 1`] = `
<div>
  <article
    class="GridItem__Wrapper-sc-qnp2w-0 kZQyAV"
    id="flyer-694041"
  >
    <div />
    <div
      class="GridItem__Description-sc-qnp2w-2 cnkKSq"
      id="flyer-description"
    >
      <span
        class="GridItem__RetailerName-sc-qnp2w-3"
        id="flyer-retailer-name"
      >
        mockName
      </span>
      <span
        class="GridItem__Title-sc-qnp2w-4 pJOSk"
        id="flyer-title"
      >
        Sport acquatici - sup gonfiabile
      </span>
      <span
        class="GridItem__CategoryName-sc-qnp2w-5 hfFSve"
        id="flyer-category-name"
      >
        mockName
      </span>
      <button
        class="GridItem__Bookmarked-sc-qnp2w-6 dtrsUa"
        data-t-flyer-is-bookmarked="true"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          style="width: 28px; height: 28px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
          />
        </svg>
      </button>
    </div>
  </article>
</div>
`;

exports[`GridItem when the grid item is rendered and is not bookmarked should have a consistent snapshot 1`] = `
<div>
  <article
    class="GridItem__Wrapper-sc-qnp2w-0 kZQyAV"
    id="flyer-694041"
  >
    <div />
    <div
      class="GridItem__Description-sc-qnp2w-2 cnkKSq"
      id="flyer-description"
    >
      <span
        class="GridItem__RetailerName-sc-qnp2w-3"
        id="flyer-retailer-name"
      >
        mockName
      </span>
      <span
        class="GridItem__Title-sc-qnp2w-4 pJOSk"
        id="flyer-title"
      >
        Sport acquatici - sup gonfiabile
      </span>
      <span
        class="GridItem__CategoryName-sc-qnp2w-5 hfFSve"
        id="flyer-category-name"
      >
        mockName
      </span>
      <button
        class="GridItem__Bookmarked-sc-qnp2w-6 foYuNw"
        data-t-flyer-is-bookmarked="false"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          style="width: 28px; height: 28px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
          />
        </svg>
      </button>
    </div>
  </article>
</div>
`;
